### PR TITLE
Add n_batch_axes option to Linear function

### LINKED ISSUE #2019: Make the number of axes for batch dimensions configurable in Linear
L.Linear accepts multiple channel dimensions. It is useful in computer vision applications where the output of the final convolution layer can be directly fed to a linear link. This feature can be seen as a flexible number of channel axes.

On the other hand, the number of batch axes is currently fixed to one, which would be cumbersome in some use cases, especially when we have to deal with a sequence of input arrays. In this case, the first two axes may be used to represent the mini-batch and the sequence. We currently need to reshape it before and after feeding it to the linear link, which has overheads both for coding and runtime.

It can be resolved by introducing an option to Linear (and also to the F.linear function) that indicates the number of batch axes.

The same discussions would also be applied to convolution links/functions.

### PR DESCRIPTION
**Pull request by mitmul**
Fix #2019

This PR enables to specify the number of batch axes when you use L.Linear.
For example, (4, 2, 3)-shaped input to L.Linear(3, 2) with a `n_batch_axes=2` argument will produce (4, 2, 2)-shaped output after this PR.

This PR actually re-sends #2265 because of tremendous conflicts after the introduction of `FunctionNode`.

### CODE REVIEW

**Comment by unnonouno** [comment]
Please remove .vscode directory.

**Comment by unnonouno** [comment]
I have a question about its interface. Do we need to add `n_batch_axes`? Is it enough to just use `x.ndim - 1` as `n_batch_axes`?

File: chainer/functions/connection/linear.py [hunk 1]
```diff
@@ -109,7 +132,13 @@ def forward(self, inputs):
                 1 in gy.shape):
             gy = numpy.ascontiguousarray(gy)
 
-        gW = gy.T.dot(x).astype(gy.dtype, copy=False)
+        if self._n_batch_axes > 1:
+            xp = cuda.get_array_module(*inputs)
+            ax = six.moves.range(self._n_batch_axes)
```
Comment (unnonouno) [inline chainer/functions/connection/linear.py]: Is it OK to use iterator here?

File: chainer/functions/connection/linear.py [hunk 2]
```diff
@@ -109,7 +132,13 @@ def forward(self, inputs):
...
+            gW = xp.tensordot(
+                gy, x, axes=(ax, ax)).astype(gy.dtype, copy=False)
```
Comment (unnonouno) [inline chainer/functions/connection/linear.py]: I think we need to use `W.dtype` here because `W` and `y` can have different dtypes.

File: chainer/functions/connection/linear.py [hunk 3]
```diff
@@ -109,7 +132,13 @@ def forward(self, inputs):
...
+        else:
+            gW = gy.T.dot(x).astype(gy.dtype, copy=False)
```
Comment (unnonouno) [inline chainer/functions/connection/linear.py]: ditto

File: chainer/functions/connection/linear.py (hunk 1 above)
Comment (mitmul) [reply chainer/functions/connection/linear.py]: Thank you for pointing out it. I checked the code below, and it works.

```
Python 3.6.4 |Anaconda, Inc.| (default, Dec 21 2017, 21:42:08)

In [2]: import numpy as np
In [3]: a = np.random.rand(1, 2, 3)
In [4]: b = np.random.rand(2, 3, 4)
In [6]: np.tensordot(a, b, axes=(range(1, 3), range(2)))
Out[6]: array([[1.77096642, 1.14276205, 1.64909544, 1.54107583]])
```

So I think it's OK to use iterator here, the documentation of numpy.tensordot just says "array_like" for the argument. Should I enclose this with `list()` or `tuple()` ?

File: chainer/functions/connection/linear.py (hunk 2 above)
Comment (mitmul) [reply chainer/functions/connection/linear.py]: Thanks! fixed it

**Comment by mitmul** [comment]
> I have a question about its interface. Do we need to add n_batch_axes? Is it enough to just use x.ndim - 1 as n_batch_axes?
I thought when the input to `Linear` is the output of a convolution whose shape is `N, C, W, H` and a user want to apply Linear to only `W, H` part, the `n_batch_axes` should be 2.

**Comment by mitmul** [comment]
@unnonouno I fixed points you pointed out

File: chainer/functions/connection/linear.py (hunk 1 above)
Comment (unnonouno) [reply chainer/functions/connection/linear.py]: Ok, thank you. You are right.

**Comment by unnonouno** [comment]
To do so, a weight matrix needs to have `(W*H, D)` shape? Your code only accepts `x.ndim` to be `n_batch_axes + 1`.
BTW it is ok to support such situation in another PR.

**Comment by unnonouno** [comment]
Check test result. It fails.

**Comment by mitmul** [comment]
@unnonouno Ah, right. The code below should be fixed.
```
x_type.ndim == self._n_batch_axes + 1,
```

I'll also check the test. Thanks

File: chainer/functions/connection/linear.py [hunk 4]
```diff
@@ -25,13 +25,13 @@ def check_type_forward(self, in_types):
         n_in = in_types.size()
         type_check.expect(2 <= n_in, n_in <= 3)
         x_type, w_type = in_types[:2]
-
+        data_size = int(numpy.prod(x_type.shape[self._n_batch_axes:]))
```
Comment (unnonouno) [inline chainer/functions/connection/linear.py]: Use `chainer.type_check.prod`

File: chainer/functions/connection/linear.py [hunk 5]
```diff
@@ -62,12 +62,19 @@ def forward(self, inputs):
                 1 in x.shape):
             x = numpy.ascontiguousarray(x)
 
-        if x.ndim != 2:
-            x = x.reshape(x.shape[:self._n_batch_axes] + (-1,))
+        if self._n_batch_axes > 1:
+            batch_shape = x.shape[:self._n_batch_axes]
+            batch_size = int(numpy.prod(batch_shape))
```
Comment (unnonouno) [inline chainer/functions/connection/linear.py]: Is this `int` required?

**Comment by mitmul** [comment]
Sorry I was really confused. So let me take the simplest solution for now.

**Comment by mitmul** [comment]
@unnonouno It's ready for re-review. Sorry for making you wait due to my fault.

**Comment by mitmul** [comment]
@unnonouno Sorry I fixed the linear.py again to reshape the output `y` again into `batch_shape + (-1,)` before returning. I think it doesn't change the shape of batch dimensions. Could you check it again?

File: tests/chainer_tests/functions_tests/connection_tests/test_linear.py [hunk 6]
```diff
@@ -74,6 +81,10 @@ def setUp(self):
 
     def forward_cpu(self, inputs):
         x, W, b = inputs
+        if self.n_batch_axes > 1:
+            batch_shape = x.shape[:self.n_batch_axes]
+            batch_size = numpy.prod(batch_shape)
+            x = x.reshape(batch_size, -1)
```
[the remark this answers is not in the document]
Comment (mitmul) [reply tests/chainer_tests/functions_tests/connection_tests/test_linear.py]: Ah, sorry. I fixed.

**Comment by unnonouno** [comment]
Can you check travis result?

**Comment by mitmul** [comment]
@unnonouno Travis tests succeeded. Could you check this PR again?

**Comment by unnonouno** [comment]
check jenkins result

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
@unnonouno The jenkins tests have succeeded.

**Comment by unnonouno** [comment]
What happens when `n_batch_axes = 0`?

**Comment by mitmul** [comment]
@unnonouno If `x.ndim` is larger than two, the input `x` is reshaped into `(x.shape[0], -1)`. Otherwise, it's treated samely as `n_batch_axes=1`. Should it raise an exception? And I noticed that it may be better to add the same interface to `L.Linear.__call__`. How do you think?

**Comment by unnonouno** [comment]
Ah, I mean there is no test case for `n_batch_axes=0` and no codes to check `n_batch_axes`. It is better to clarify the condition about range of `n_batch_axes` and to add test cases for boundary condition.

Yes, we need to add the same feature to the link. But you can split the PR in another one.

**Comment by mitmul** [comment]
I see. I added a value check for `n_batch_axes` and test cases for the boundary condition.

**Comment by unnonouno** [comment]
I see. That sounds good.

**Comment by unnonouno** [comment]
LGTM except flake8

**Comment by mitmul** [comment]
@unnonouno Sorry for late reaction. I fixed the flake8 errors.

**Comment by mitmul** [comment]
Jenkins, test this please. You can do it.

**Comment by unnonouno** [comment]
jenkins test this please

**Comment by unnonouno** [comment]
LGTM
